tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -778,6 +778,79 @@ private static String extractAndTransformJavaScript(Document document, Webpath j
     return scriptContent;
   }
 
+  private static void cloneChildrenWithoutWhitespace(Element src, Element dest) {
+    List<Node> toMove = new ArrayList<Node>();
+    for (Node node : src.childNodes()) {
+      if (node instanceof TextNode && ((TextNode) node).isBlank()) {
+        continue;
+      }
+      toMove.add(node);
+    }
+    for (int i = 0; i < toMove.size(); i++) {
+      Node node = toMove.get(i);
+      dest.appendChild(node.clone());
+    }
+  }
+
+  // When we inline the HTML based on `<link rel="import">` in `transform`, we
+  // replace the link element with parsed document. This makes us have nested
+  // documents and jsoup's Node.outerHtml (or Node.toString) are incapable of
+  // properly outputting that. Here, we flatten the document by combining all
+  // elements in `<head>` and `<body>` of nested document in one `<head>` and
+  // `<body>`.
+  //
+  // Examples:
+  // // Input
+  // <#root> <!-- document -->
+  //   <html>
+  //     <head><#root><html><body>foo</body></html></#root></head>
+  //     <body><span>bar</span></body>
+  //   </html>
+  // </html>
+  // // Output
+  // <#root> <!-- document -->
+  //   <html>
+  //     <head></head>
+  //     <body>foo<span>bar</span></body>
+  //   </html>
+  // </html>
+  private static Document getFlattenedDocument(Document document) {
+    Document flatDoc = new Document("/");
+    flatDoc.normalise();
+    Element rootDocumentHead = flatDoc.head();
+    Element rootDocumentBody = flatDoc.body();
+
+    Node currentNode = document;
+    while (currentNode != null) {
+      // Do not clone the element if it is a `document` inside `<head>`.
+      // We want to traverse further and get all elements from `<head>` and `<body>`
+      if (currentNode.parentNode() != null && currentNode.parentNode().nodeName().equals("head")
+          && !(currentNode instanceof Document)) {
+        rootDocumentHead.appendChild(currentNode.clone());
+      }
+
+      if (currentNode.nodeName().equals("body")) {
```

Comment:
Is it possible to stop traversing deeper into body elements or do we have body elements nested inside body elements?  If the latter (nested body elements) then that's probably worth mentioning in the doc.